packages/test/test-end-to-end-tests/src/test/localTestSignals.spec.ts

```diff
@@ -7,29 +7,92 @@ import { strict as assert } from "assert";
 
 import { describeCompat } from "@fluid-private/test-version-utils";
 import { ConnectionState } from "@fluidframework/container-loader";
-import { IInboundSignalMessage } from "@fluidframework/runtime-definitions/internal";
+import { IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions/internal";
+import type {
+	IContainerRuntimeBase,
+	IInboundSignalMessage,
+} from "@fluidframework/runtime-definitions/internal";
 import {
 	DataObjectFactoryType,
 	ITestContainerConfig,
 	ITestFluidObject,
 	ITestObjectProvider,
 	getContainerEntryPointBackCompat,
-	timeoutPromise,
+	timeoutPromise as timeoutPromiseUnnamed,
+	TimeoutWithError,
+	TimeoutWithValue,
 } from "@fluidframework/test-utils/internal";
 
+type IContainerRuntimeBaseWithClientId = IContainerRuntimeBase & {
+	clientId?: string | undefined;
+};
+
+enum RuntimeLayer {
+	dataStoreRuntime = "dataStoreRuntime",
+	containerRuntime = "containerRuntime",
+}
+
+interface SignalClient {
+	dataStoreRuntime: IFluidDataStoreRuntime;
+	containerRuntime: IContainerRuntimeBaseWithClientId;
+	signalReceivedCount: number;
+	clientId: string | undefined;
+}
+
 const testContainerConfig: ITestContainerConfig = {
 	fluidDataObjectType: DataObjectFactoryType.Test,
 };
 
-const waitForSignal = async (...signallers: { once(e: "signal", l: () => void): void }[]) =>
-	Promise.all(
+async function timeoutPromise<T = void>(
+	executor: (controller: {
+		resolve: (value: T | PromiseLike<T>) => void;
+		reject: (reason?: any) => void;
+	}) => void,
+	timeoutOptions: TimeoutWithError | TimeoutWithValue<T> = {},
+): Promise<T> {
+	return timeoutPromiseUnnamed(
+		(resolve, reject) => executor({ resolve, reject }),
+		timeoutOptions,
+	);
+}
+
+async function waitForSignal(...signallers: { once(e: "signal", l: () => void): void }[]) {
+	return Promise.all(
 		signallers.map(async (signaller, index) =>
-			timeoutPromise((resolve) => signaller.once("signal", () => resolve()), {
+			timeoutPromise((controller) => signaller.once("signal", () => controller.resolve()), {
```

Comment:
Just FYI, another way to write this is
```typescript
timeoutPromise(({resolve}) => signaller.once("signal", () => resolve()), {
```
It has advantages of (1) changing less code, if we do end up cleaning up the uses and (2) making it clear what is used. In the common case `resolve` is used. So, in the case where just `reject` is used, seeing that is helpful.
(Also `controller` is a very nebulous term generally; so not using is a plus.